File: src/mev/sensors/MempoolCongestionSensor.ts
/**
 * MempoolCongestionSensor - Real-time mempool congestion score
 * Ported from AxionCitadel's mev_risk_arb/sensors/mempool_congestion.py
 * 
 * Calculates a 0-1 congestion score based on:
 * 1. Pending transactions ratio
 * 2. Gas usage deviation
 * 3. Base fee velocity (EIP-1559 dynamics)
 */

import { ethers } ,Provider } from 'ethers';
import { SensorReading } from '../types/TransactionType';

export interface CongestionWeights {
  pendingRatio: number;
  gasDeviation: number;
  feeVelocity: number;
}

const DEFAULT_WEIGHTS: CongestionWeights = {
  pendingRatio: 0.4,
  gasDeviation: 0.3,
  feeVelocity: 0.3,
};

export class MempoolCongestionSensor {
  private provider: Provider;
  private weights: CongestionWeights;
  private windowSize: number;

  constructor(
    provider: Provider,
    windowSize: number = 5,
    weights?: Partial<CongestionWeights>
  ) {
    this.provider = provider;
    this.windowSize = windowSize;
    this.weights = { ...DEFAULT_WEIGHTS, ...weights };
  }

  /**
   * Get real-time mempool congestion score (0-1 scale)
   */
  async getCongestionScore(): Promise<number> {
    try {
      // 1. Pending transactions ratio
      const pendingRatio = await this.calculatePendingRatio();

      // 2. Gas usage deviation
      const gasDeviation = await this.calculateGasDeviation();

      // 3. Base fee velocity (EIP-1559 dynamics)
      const feeVelocity = await this.calculateFeeVelocity();

      // Weighted composition
      const congestionScore =
        this.weights.pendingRatio * pendingRatio +
        this.weights.gasDeviation * Math.min(gasDeviation, 1.0) +
        this.weights.feeVelocity * Math.min(Math.abs(feeVelocity), 1.0);

      return congestionScore;
    } catch (error) {
      console.error('Error in getCongestionScore:', error);
      // Fallback to neutral score
      return 0.5;
    }
  }

  /**
   * Get sensor reading with metadata
   */
  async getReading(): Promise<SensorReading> {
    const value = await this.getCongestionScore();
    return {
      value,
      timestamp: Date.now(),
      source: 'MempoolCongestionSensor',
    };
  }

  /**
   * Calculate pending transaction ratio
   */
  private async calculatePendingRatio(): Promise<number> {
    try {
      // Get pending block with full transactions
      const pendingBlock = await this.provider.getBlock('pending');
      const pendingCount = pendingBlock?.transactions?.length || 0;

      // Get latest block to estimate capacity
      const latestBlock = await this.provider.getBlock('latest');
      const txPerBlock = latestBlock?.transactions?.length || 1;

      // Cap at 10 blocks equivalent
      const pendingRatio = Math.min(pendingCount / Math.max(txPerBlock * 10, 1), 1.0);

      return pendingRatio;
    } catch (error) {
      console.warn('Error calculating pending ratio:', error);
      return 0;
    }
  }

  /**
   * Calculate gas usage deviation across recent blocks
   */
  private async calculateGasDeviation(): Promise<number> {
    try {
      const currentBlockNumber = await this.provider.getBlockNumber();
      const blocks: ethers.providers.Block[] = [];

      // Fetch recent blocks
      for (let i = 0; i < this.windowSize; i++) {
        if (currentBlockNumber - i >= 0) {
          const block = await this.provider.getBlock(currentBlockNumber - i);
          if (block) {
            blocks.push(block);
          }
        }
      }

      if (blocks.length === 0) {
        return 0;
      }

      // Calculate gas usage ratios
      const gasRatios = blocks
        .filter((b) => b.gasLimit > 0n)
        .map((b) => Number(b.gasUsed) / Number(b.gasLimit));

      if (gasRatios.length < 2) {
        return 0;
      }

      // Calculate standard deviation
      const mean = gasRatios.reduce((a, b) => a + b, 0) / gasRatios.length;
      const squaredDiffs = gasRatios.map((r) => Math.pow(r - mean, 2));
      const variance = squaredDiffs.reduce((a, b) => a + b, 0) / gasRatios.length;
      const stdev = Math.sqrt(variance);

      // Amplify variance sensitivity
      return stdev * 2;
    } catch (error) {
      console.warn('Error calculating gas deviation:', error);
      return 0;
    }
  }

  /**
   * Calculate base fee velocity (rate of change)
   */
  private async calculateFeeVelocity(): Promise<number> {
    try {
      const currentBlockNumber = await this.provider.getBlockNumber();
      const blocks: ethers.providers.Block[] = [];

      // Fetch recent blocks
      for (let i = 0; i < this.windowSize; i++) {
        if (currentBlockNumber - i >= 0) {
          const block = await this.provider.getBlock(currentBlockNumber - i);
          if (block) {
            blocks.push(block);
          }
        }
      }

      // Extract base fees (EIP-1559)
      const baseFees = blocks
        .filter((b) => b.baseFeePerGas !== undefined && b.baseFeePerGas !== null)
        .map((b) => b.baseFeePerGas!.toNumber());

      if (baseFees.length < 2) {
        return 0;
      }

      // Calculate percentage change
      const lastFee = baseFees[baseFees.length - 1];
      const firstFee = baseFees[0];

      if (lastFee === 0) {
        return firstFee === 0 ? 0 : 1.0;
      }

      const feeVelocity = (firstFee - lastFee) / lastFee;
      return feeVelocity;
    } catch (error) {
      console.warn('Error calculating fee velocity:', error);
      return 0;
    }
  }
}

File: src/mev/sensors/mempool_congestion_sensor.py
"""
Mempool Congestion Sensor - Real-time MEV Risk Monitoring

Extracted from AxionCitadel - Operation First Light validated
Monitors mempool congestion to adjust MEV risk calculations dynamically

Metrics:
1. Pending transactions ratio (mempool depth)
2. Gas usage deviation (block fullness variance)
3. Base fee velocity (EIP-1559 dynamics)

Returns: Congestion score (0.0 - 1.0)
"""

from web3 import Web3
import statistics
from typing import Tuple, List


class MempoolCongestionSensor:
    """Real-time mempool congestion monitoring for MEV risk adjustment"""
    
    def __init__(self, rpc_url: str, window_size: int = 5, weights: Tuple[float, float, float] = (0.4, 0.3, 0.3)):
        """\n        Initialize mempool congestion sensor\n        
        Args:\n            rpc_url: Web3 RPC endpoint URL (e.g., Arbitrum, Ethereum)\n            window_size: Number of blocks to analyze for historical trends\n            weights: Tuple of (pending_ratio_weight, gas_deviation_weight, fee_velocity_weight)\n        """
        self.w3 = Web3(Web3.HTTPProvider(rpc_url))
        self.window_size = window_size
        self.weights = weights
        
        if not self.w3.is_connected():
            raise ConnectionError(f"Failed to connect to RPC endpoint: {rpc_url}")
    
    def get_congestion_score(self) -> float:
        """\n        Calculate real-time mempool congestion score\n        
        Returns:\n            Congestion score (0.0 = empty, 1.0 = extremely congested)\n        """
        try:
            pending_ratio = self._calculate_pending_ratio() 
            gas_deviation = self._calculate_gas_deviation() 
            fee_velocity = self._calculate_fee_velocity() 
            
            congestion_score = (
                self.weights[0] * pending_ratio +
                self.weights[1] * min(gas_deviation, 1.0) +
                self.weights[2] * min(abs(fee_velocity), 1.0)
            )
            
            return min(congestion_score, 1.0)
            
        except Exception as e:
            print(f"Error in get_congestion_score: {e}")
            return 0.5
    
    def _calculate_pending_ratio(self) -> float:
        """Calculate ratio of pending transactions to block capacity"""
        try:
            pending_block = self.w3.eth.get_block('pending', full_transactions=True)
            pending_count = len(pending_block.transactions)
            
            latest_block = self.w3.eth.get_block('latest')
            tx_per_block = len(latest_block.transactions) if latest_block.transactions else 1
            
            pending_ratio = min(pending_count / max(tx_per_block * 10, 1), 1.0)
            return pending_ratio
            
        except Exception:
            return 0.0
    
    def _calculate_gas_deviation(self) -> float:
        """Calculate gas usage deviation across recent blocks"""
        try:
            blocks = self._fetch_recent_blocks()
            
            if not blocks:
                return 0.0
            
            gas_ratios = [b.gasUsed / b.gasLimit for b in blocks if b.gasLimit > 0]
            
            if len(gas_ratios) < 2:
                return 0.0
            
            gas_deviation = statistics.stdev(gas_ratios) * 2
            return gas_deviation
            
        except Exception:
            return 0.0
    
    def _calculate_fee_velocity(self) -> float:
        """Calculate base fee velocity (rate of change)"""
        try:
            blocks = self._fetch_recent_blocks()
            base_fees = [b.baseFeePerGas for b in blocks if b.baseFeePerGas is not None]
            
            if len(base_fees) < 2:
                return 0.0
            
            if base_fees[-1] > 0:
                fee_velocity = (base_fees[0] - base_fees[-1]) / base_fees[-1]
            else:
                fee_velocity = 0.0 if base_fees[0] == 0 else 1.0
            
            return fee_velocity
            
        except Exception:
            return 0.0
    
    def _fetch_recent_blocks(self) -> List:
        """Fetch recent blocks for analysis"""
        blocks = []
        current_block_number = self.w3.eth.block_number
        
        for i in range(self.window_size):
            if current_block_number - i >= 0:
                block = self.w3.eth.get_block(current_block_number - i)
                blocks.append(block)
            else:
                break
        
        return blocks
    
    def get_metrics(self) -> dict:
        """Get detailed congestion metrics for debugging/monitoring"""
        return {
            "congestion_score": self.get_congestion_score(),
            "pending_ratio": self._calculate_pending_ratio(),
            "gas_deviation": self._calculate_gas_deviation(),
            "fee_velocity": self._calculate_fee_velocity(),
            "window_size": self.window_size,
            "weights": self.weights
        }